Virtual thread reference documentation and resteasy reactive how-to (#34913)
quarkusio/quarkus: docs/src/main/asciidoc/virtual-threads.adoc

= Writing simpler reactive REST services with Quarkus Virtual Thread support

include::_attributes.adoc[]
:diataxis-type: reference
:resteasy-reactive-api: https://javadoc.io/doc/io.quarkus.resteasy.reactive/resteasy-reactive/{quarkus-version}
:runonvthread: https://javadoc.io/doc/io.smallrye.common/smallrye-common-annotation/latest/io/smallrye/common/annotation/RunOnVirtualThread.html
:blockingannotation: https://javadoc.io/doc/io.smallrye.common/smallrye-common-annotation/latest/io/smallrye/common/annotation/Blocking.html
:vthreadjep: https://openjdk.org/jeps/425
:thread: https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/Thread.html
:pgsql-driver: https://javadoc.io/doc/org.postgresql/postgresql/latest/index.html

This guide explains how to benefit from Java 19+ virtual threads in Quarkus application.

== What are virtual threads?

=== Terminology
OS thread::
A "thread-like" data structure managed by the Operating System.

Platform thread::
Until Java 19, every instance of the link:{thread}[Thread] class was a platform thread, a wrapper around an OS thread.

> Creating a platform thread creates an OS thread, and blocking a platform thread, consequently, blocks an OS thread.

Reviewer: ```suggestion
Creating a platform thread creates an OS thread, and blocking a platform thread blocks an OS thread.
```

> Virtual thread::

Reviewer: We might want to add a reference to https://dl.acm.org/doi/10.1145/3583678.3596895

Their state is stored in the heap, and another virtual thread is executed on the same Java platform (carrier) thread.

=== The Continuation Dance
As mentioned above, the JVM schedules the virtual threads. These virtual threads are mounted on carrier threads. The scheduling comes with a pinch of magic. When the virtual thread attempts to use blocking I/O, the JVM _transforms_ this call into a non-blocking one, unmounts the virtual thread, and mounts another virtual thread on the carrier thread. When the I/O completes, the _waiting_ virtual thread becomes eligible again and will be re-mounted on a carrier thread to continue its execution. For the user, all this dance is invisible. Your synchronous code is executed asynchronously.

Note that the virtual thread may not be re-mounted on the same carrier thread.

[[cpu-bound]]
=== Virtual threads are useful for I/O-bound workloads only
We now know we can create more virtual threads than platform threads. One could be tempted to use virtual threads to perform long computations (CPU-bound workload).
It is useless and counterproductive.

> CPU-bound doesn't consist in quickly swapping threads while they need to wait for the completion of an I/O, but in leaving

Reviewer 2: ```suggestion
CPU-bound doesn't consist of quickly swapping threads while they need to wait for the completion of an I/O, but in leaving
```

them attached to a CPU core to compute something.

> In this scenario, it is useless to have thousands of threads if we have tens of CPU cores, virtual threads won't enhance the performance of CPU-bound workloads.

Reviewer 2: ```suggestion
In this scenario, it is worse than useless to have thousands of threads if we have tens of CPU cores, virtual threads won't enhance the performance of CPU-bound workloads.
```

Even worse, when running a CPU-bound workload on a virtual thread, the virtual thread monopolizes the carrier thread on which it is mounted.
It will either reduce the chance for the other virtual thread to run or will start creating new carrier threads, leading to high memory usage.

== Run code on virtual threads using @RunOnVirtualThread

In Quarkus, the support of virtual thread is implemented using the link:{runonvthread}[@RunOnVirtualThread] annotation.
This section briefly overviews the rationale and how to use it.
There are dedicated guides for extensions supporting that annotation, such as // TODO.

[[why-not]]
=== Why not run everything on virtual threads?
As mentioned above, not everything can run safely on virtual threads.
The risk of **monopolization** can lead to high-memory usage.
Also, there are situations where the virtual thread cannot be unmounted from the carrier thread.
This is called **pinning**.
Finally, some libraries use `ThreadLocal` to store and reuse objects.
Using virtual threads with these libraries will lead to massive allocation, as the intentionally pooled objects will be instantiated for every (disposable and generally short-lived) virtual thread.

> As of today, it is not possible to use virtual threads unthinkingly.

Reviewer 2: ```suggestion
As of today, it is not possible to use virtual threads in a carefree manner.
```

> It could lead to memory and resource starvation issues.

Reviewer 2: ```suggestion
Following such a laissez-faire approach could quickly lead to memory and resource starvation issues.
```

> Thus, Quarkus uses an explicit model until the mentioned issues disappear.

Reviewer 2: ```suggestion
Thus, Quarkus uses an explicit model until the aformentioned issues disappear (as the Java ecosystem matures).
```

It is also the reason why _reactive_ extensions have the virtual thread support, and rarely the _classic_ ones.
We need to know when to dispatch on a virtual thread.

> It is essential to understand that these issues are not  Quarkus limitations or bugs but are due to the Java ecosystem that would need to evolve to become virtual thread friendly.

Reviewer 2: ```suggestion
It is essential to understand that these issues are not  Quarkus limitations or bugs but are due to the current stare Java ecosystem which needs to evolve to become virtual thread friendly.
```

[[monopolization]]
==== Monopolization cases
The monopolization has been explained in the xref:cpu-bound[Virtual threads are useful for I/O-bound workloads only] section.
When running long computations, we do not allow the JVM to unmount and switch to another virtual thread until the virtual thread terminates.
Indeed, the current scheduler does not support preempting tasks.

This monopolization can lead to the creation of new carrier threads to take care of the other virtual threads.
Creating carrier threads results in creating platform threads.
So, there is a memory cost associated with this creation.

Suppose you run in a constrained environment, such as containers. In that case, monopolization can quickly become a concern, as the high memory usage can lead to out-of-memory issues and container termination.
The memory usage may be higher than with regular worker threads because of the inherent cost of the scheduling and virtual threads.

[[pinning]]
==== Pinning cases

> The notion of "cheap blocking" might not always be true: a virtual thread might _pin_ its carrier on certain occasions.

Reviewer 2: ```suggestion
The promise of "cheap blocking" might not always hold: a virtual thread might _pin_ its carrier on certain occasions.
```

The platform thread is blocked in this situation, precisely as it would have been in a typical blocking scenario.

According to link:{vthreadjep}[JEP 425] this can happen in two situations:

- when a virtual thread performs a blocking operation inside a `synchronized` block or method
- when it executes a blocking operation inside a native method or a foreign function

It can be reasonably easy to avoid these situations in your code, but verifying every dependency you use is hard.
Typically, while experimenting with virtual threads, we realized that old versions of the link:{pgsql-driver}[postgresql-JDBC driver]
results in frequent pinning.
Most JDBC drivers still pin the carrier thread.
Even worse, lots of widespread libraries are pinning and would require code changes.

[[pooling]]
==== The pooling case

> Some libraries are using `ThreadLocal` as a polling mechanism.

Reviewer 2: ```suggestion
Some libraries are using `ThreadLocal` as a pooling mechanism.
```

> Libraries like Jackson or Netty suppose that the application uses a limited number of threads, which are recycled (using a thread pool) to run multiple (unrelated) tasks.

Reviewer 2: ```suggestion
Extremely popular iibraries like Jackson and Netty assume that the application uses a limited number of threads, which are recycled (using a thread pool) to run multiple (unrelated) tasks.
```

    @Path("/virtual")
    @RunOnVirtualThread
    public Fortune virtualThread() {
        // Runs on a virtual thread
        var list = repository.findAllAsyncAndAwait();
        return pickOne(list);
    }

}
----

The following table summarizes the options:

|===
|Model |Example of signature |Pros |Cons

|Synchronous code on worker thread
|`Fortune blocking()`
|Simple code
|Use worker thread (limit concurrency)

|Reactive code on event loop
|`Uni<Fortune> reactive()`
|High concurrency and low resource usage
|More complex code

|Synchronous code on virtual thread
|`@RunOnVirtualThread Fortune vt()`
|Simple code
|Risk of pinning, monopolization and under-efficient object pooling
|===

Note that all three models can be used in a single application.

=== Run application using virtual threads

> Until Java 21, virtual threads are still an experimental feature, you need to start your application with the `--enable-preview` flag:

Reviewer 2: ```suggestion
Prior to Java 21, virtual threads are still an experimental feature, you need to start your application with the `--enable-preview` flag:
```

|More complex code

|Synchronous code on virtual thread
|`@RunOnVirtualThread Fortune vt()`
|Simple code
|Risk of pinning, monopolization and under-efficient object pooling
|===

Note that all three models can be used in a single application.

=== Run application using virtual threads

Prior to Java 21, virtual threads are still an experimental feature, you need to start your application with the `--enable-preview` flag:

[source, bash]
----
java --enable-preview -jar target/quarkus-app/quarkus-run.jar
----

== Use virtual thread friendly clients

As mentioned in the href:why-not[Why not run everything on virtual threads?] section, the Java ecosystem is not entirely ready for virtual threads.
So, you need to be careful, especially when using a libraries doing I/O.

Fortunately, Quarkus provides a massive ecosystem that is ready to be used in virtual threads.
Mutiny, the reactive programming library used in Quarkus, and the Vert.x Mutiny bindings provides the ability to write blocking code (so, no fear, no learning curve) which do not pin the carrier thread.

As a result:

1. Quarkus extensions providing blocking APIs on top of reactive APIs can be used in virtual threads.

> This includes the reactive rest client, the redis data source...

Author: mailer!

Note that, to use in-container build, you must have Docker or Podman installed on your machine.

Then, add to your `application.properties` file:

[source, properties]
----
quarkus.native.additional-build-args=--enable-preview <1>

# In-container build to get a linux 64 executable
quarkus.native.container-build=true <2>
quarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-20 <3>
----
<1> The `enable-preview` flag in only necessary until Java 21.
<2> Enables the in-container build
<3> The builder container to use. Make sure it supports virtual threads

[IMPORTANT]
.From ARM/64 to AMD/64
====
If you are using a Mac M1 or M2 (using an ARM64 CPU), you need to be aware that the native executable you will get using an in-container build will be a Linux executable, but using your host (ARM 64) architecture.
You can use emulation to force the architecture when using Docker with the following property:

[source, properties]
----
quarkus.native.container-runtime-options=--platform=linux/amd64
----

Be aware that it increases the compilation time... a lot (>10 minutes).
====

> == Containerize applications using virtual threads

Reviewer 3: Shall we move this section before [[native]] ? 
In this sequence it does native compilation, JVM container, native container. I think it would be easier to follow if it was JVM container, native compilation, native container.

Author: so splitting the containerization section into two distinguished section.

Reviewer 3: When I was reading the containerization section I thought it was about containerization of native apps.

Author: Done
